integrations/terraform-cloud/client.py

```diff
@@ -0,0 +1,286 @@
+import httpx
+from typing import List, Dict, Any, AsyncGenerator
+from loguru import logger
+from port_ocean.context.event import event
+from enum import StrEnum
+
+
+TERRAFORM_WEBHOOK_EVENTS = ['run:applying',
+                            'run:completed',
+                            'run:created', 
+                            'run:errored', 
+                            'run:needs_attention',
+                            'run:planning']
+
+class CacheKeys(StrEnum):
+    WORKSPACES = "workspace"
+    RUNS = "run"
+    STATE_VERSION_OUTPUTS = "state-version-output"
+    STATE_VERSION = "state-version"
+
+
+PAGE_SIZE = 100
+
+class TerraformClient:
+
+    def __init__(self, terraform_base_url: str, auth_token: str) -> None:
+        self.terraform_base_url = terraform_base_url
+        self.base_headers = {"Authorization": f"Bearer {auth_token}", "Content-Type": "application/vnd.api+json"}
+        self.api_url = f"{self.terraform_base_url}/api/v2"
+        self.client = httpx.AsyncClient(headers=self.base_headers)
+
+    async def get_single_workspace(self, workspace_id: str) -> Dict[str, Any]:
+        
+        try:
+            workspace_url = f"{self.api_url}/workspaces/{workspace_id}"
+            response = await self.client.get(workspace_url)
+            response.raise_for_status()
+            return response.json().get('data', {})
+        
+        except httpx.HTTPStatusError as e:
+            logger.error(f"HTTP error with status code : {e.response.status_code} and response text: {e.response.text}")
+        
+        except httpx.HTTPError as e:
+            logger.error(f"An error occured while fetching terraform data: {e}")
+
+
+    async def get_paginated_workspaces(self) -> List[Dict[str, any]]:
+
+        cache_key = CacheKeys.WORKSPACES
+
+        if cache := event.attributes.get(cache_key):
+            logger.info("Fetching paginated workspaces from cache")
+            yield cache
+            return
+
+        all_workspaces = []
+
+        try:        
+            page = 1
+
+            logger.info("Getting workspaces from Terraform")
+            while True:
+
+                async for organization in self.get_organizations():
+
+                    workspace_response = await self.client.get(
+                        f"{self.api_url}/organizations/{organization['id']}/workspaces?page[number]={page}&page[size]={PAGE_SIZE}"
+                    )
+                    workspace_response.raise_for_status()
+                    workspaces_data = workspace_response.json()
+                    workspaces = workspaces_data.get('data', [])
+
+                    for workspace in workspaces:
+                        workspace['__currentStateVersion'] = self.get_current_state_version(workspace['id'])

```

Comment:
why is there a need for this

i their docs it says the reponse should hold information about the current state version

Reply:
That's actually true.

relationships.current-state-version.data.id

```json
            "relationships": {
                "organization": {
                    "data": {
                        "id": "example-org-162af6",
                        "type": "organizations"
                    }
                },
                "current-run": {
                    "data": {
                        "id": "run-jcPBr7DTcNMhs54U",
                        "type": "runs"
                    },
                    "links": {
                        "related": "/api/v2/runs/run-jcPBr7DTcNMhs54U"
                    }
                },
                "latest-run": {
                    "data": {
                        "id": "run-jcPBr7DTcNMhs54U",
                        "type": "runs"
                    },
                    "links": {
                        "related": "/api/v2/runs/run-jcPBr7DTcNMhs54U"
                    }
                },
                "outputs": {
                    "data": [],
                    "links": {
                        "related": "/api/v2/workspaces/ws-pN259iR1J3cW8ivr/current-state-version-outputs"
                    }
                },
                "remote-state-consumers": {
                    "links": {
                        "related": "/api/v2/workspaces/ws-pN259iR1J3cW8ivr/relationships/remote-state-consumers"
                    }
                },
                "current-state-version": {
                    "data": {
                        "id": "sv-auvLUF9XRyp4UV6W",
                        "type": "state-versions"
                    },
                    "links": {
                        "related": "/api/v2/workspaces/ws-pN259iR1J3cW8ivr/current-state-version"
                    }
                },
                "current-configuration-version": {
                    "data": {
                        "id": "cv-khh57PpYRnh9PQVK",
                        "type": "configuration-versions"
                    },
                    "links": {
                        "related": "/api/v2/configuration-versions/cv-khh57PpYRnh9PQVK"
                    }
                },
                "agent-pool": {
                    "data": null
                },
                "readme": {
                    "data": null
                },
                "project": {
                    "data": {
                        "id": "prj-wnLLjhXa3XArrRFR",
                        "type": "projects"
                    }
                },
                "current-assessment-result": {
                    "data": null
                },
                "vars": {
                    "data": []
                }
            }
```

Reply:
Got it